### ISSUE TITLE
[Bug]: compact wrong: download from blob stroage wrong

### ISSUE THREAD

**Issue by yanliang567** [milvus-io/milvus#24901]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: master-20230614-56380ea7
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka):    pulsar
- SDK version(e.g. pymilvus v2.0.0rc2):
```

### Current Behavior

cpmact wrong error after upgrading to master-20230614-56380ea7,
`[2023/06/15 01:34:02.234 +00:00] [WARN] [datanode/compactor.go:714] ["compact wrong"] [planID=442182520523456940] [error="download from blob storage wrong"] [errorVerbose="download from blob storage wrong\n(1) attached stack trace\n  -- stack trace:\n  | github.com/milvus-io/milvus/internal/datanode.init\n  | \t/go/src/github.com/milvus-io/milvus/internal/datanode/binlog_io.go:40\n  | runtime.doInit\n  | \t/usr/local/go/src/runtime/proc.go:6222\n  | runtime.doInit\n  | \t/usr/local/go/src/runtime/proc.go:6199\n  | runtime.doInit\n  | \t/usr/local/go/src/runtime/proc.go:6199\n  | runtime.doInit\n  | \t/usr/local/go/src/runtime/proc.go:6199\n  | runtime.doInit\n  | \t/usr/local/go/src/runtime/proc.go:6199\n  | runtime.doInit\n  | \t/usr/local/go/src/runtime/proc.go:6199\n  | runtime.main\n  | \t/usr/local/go/src/runtime/proc.go:233\n  | runtime.goexit\n  | \t/usr/local/go/src/runtime/asm_amd64.s:1571\nWraps: (2) download from blob storage wrong\nError types: (1) *withstack.withStack (2) *errutil.leafError"]`

### Expected Behavior

compaction succeed

### Steps To Reproduce

```markdown
1. my collection has close to 100m-128d vectors, and has been running for a few weeks
2. upgraded to master-20230614-56380ea7
3. compaction error, and cannot load the collection again.

One more thing is that only one of data node is working for compaction after upgrading, which I think, is also an issue.
```

### Milvus Log

Pod names:
yanliang-cd-mmszh-milvus-datanode-5944664c88-m6jzr                1/1     Running       0                10m     10.102.9.58     devops-node13   <none>           <none>
yanliang-cd-mmszh-milvus-datanode-5944664c88-vjpcq                1/1     Running       0                10m     10.102.6.244    devops-node10   <none>           <none>
yanliang-cd-mmszh-milvus-indexnode-5479674469-2p2kl               1/1     Running       0                10m     10.102.6.243    devops-node10   <none>           <none>
yanliang-cd-mmszh-milvus-indexnode-5479674469-8zbdm               1/1     Running       0                10m     10.102.9.67     devops-node13   <none>           <none>
yanliang-cd-mmszh-milvus-mixcoord-74f5cf8bc8-cnz5q                1/1     Running       0                10m     10.102.9.76     devops-node13   <none>           <none>
yanliang-cd-mmszh-milvus-proxy-76b88f59fb-zxzsn                   1/1     Running       0                10m     10.102.9.62     devops-node13   <none>           <none>
yanliang-cd-mmszh-milvus-querynode-594d949b7d-2mzmx               1/1     Running       0                10m     10.102.9.60     devops-node13   <none>           <none>
yanliang-cd-mmszh-milvus-querynode-594d949b7d-7lvv8               1/1     Running       0                10m     10.102.7.179    devops-node11   <none>           <none>
yanliang-cd-mmszh-milvus-querynode-594d949b7d-grc4c               1/1     Running       0                10m     10.102.6.242    devops-node10   <none>           <none>
yanliang-cd-mmszh-milvus-querynode-594d949b7d-pgml2               1/1     Running       0                10m     10.102.10.27    devops-node20   <none>           <none>
yanliang-cd-mmszh-milvus-querynode-594d949b7d-pkf7f               1/1     Running       0                10m     10.102.9.96     devops-node13   <none>           <none>

### Anything else?

only one of the data node is working after upgrading is also an issue.
![image](https://github.com/milvus-io/milvus/assets/82361606/334f4d1a-3c81-47e7-992b-333f2d2f6986)

**Comment by congqixia** [comment]
![image](https://github.com/milvus-io/milvus/assets/84113973/c8ec7906-9c2b-4180-89df-4363c1b79b78)
compaction failed due to download timeout

**Comment by yanliang567** [comment]
so why it failed in download?

**Comment by congqixia** [comment]
in brief: timeout

> so why it failed in download?

**Comment by bedivere2018** [comment]
`datanode [2023/07/18 20:25:07.740 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411090010] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c  │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"]
datanode [2023/07/18 20:30:32.962 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411091951] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c  │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"]
datanode [2023/07/18 20:36:36.685 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411094241] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c  │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"]
datanode [2023/07/18 20:42:32.962 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411096332] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c  │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"]
datanode [2023/07/18 20:48:36.279 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411098495] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c  │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"]
datanode [2023/07/18 20:54:32.963 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411100568] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c  │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"]
`

I have also encountered the same problem, and it has been bothering me for a long time. Is there any progress?

**Comment by xiaofan-luan** [comment]
> milvus 2.2.11
> 
> `datanode [2023/07/18 20:25:07.740 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411090010] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/18 20:30:32.962 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411091951] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/18 20:36:36.685 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411094241] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/18 20:42:32.962 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411096332] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/18 20:48:36.279 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411098495] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/18 20:54:32.963 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411100568] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] `
> 
> I have also encountered the same problem, and it has been bothering me for a long time. Is there any progress?

which storage did you use? For S3 could be caused by S3 quota limit?
If you do very frequent insert/delete with flush operation this could happen

**Comment by bedivere2018** [comment]
> > milvus 2.2.11
> > `datanode [2023/07/18 20:25:07.740 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411090010] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/18 20:30:32.962 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411091951] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/18 20:36:36.685 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411094241] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/18 20:42:32.962 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411096332] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/18 20:48:36.279 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411098495] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/18 20:54:32.963 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861411100568] [error="download from blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.c │om/milvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] `
> > I have also encountered the same problem, and it has been bothering me for a long time. Is there any progress?
> 
> which storage did you use? For S3 could be caused by S3 quota limit? If you do very frequent insert/delete with flush operation this could happen

@xiaofan-luan 
A minio cluster consisting of 5 servers is used. 
I will try removing all the 'flush' statements in the code and continue observing. 
When running concurrently, there is indeed a possibility of frequent flushing. Thank you for your reminder.

**Comment by bedivere2018** [comment]
@xiaofan-luan 
After removing "flush", there is indeed no more "compact wrong" log. However, I have encountered two problems now:
1. The number of segments continues to increase.
2. After deleting records, the entity count does not decrease.
Do these problems have any relation to the previous "compact wrong"? Could it be causing the compact process to get stuck? Could you help me figure out what I should do now?

**Comment by xiaofan-luan** [comment]
> 

Delete records will be removed after compaction.
This happens if: 
1) if segment owns enough deletes
2) trigger compaction manually
3) if there are small files

**Comment by bedivere2018** [comment]
@xiaofan-luan Thank you. I'm  trying to trigger compaction manually.
In pymilvus 2.2.13, when I call collection.compact(), I found that it only lasts for about 5 minutes each time. 
Does this mean that I need to call compact repeatedly? Can a single collection be called concurrently? Also, can I check the status of compact?

**Comment by xiaofan-luan** [comment]
> 

compact is actually happened in the back ground?
@XuanYang-cn any idea?

**Comment by bedivere2018** [comment]
@xiaofan-luan 
I haven't installed Prometheus yet, so I can only rely on basic server monitoring. 
I have made a few calls and noticed some fluctuations in CPU and disk IO, but they were not very high. 
The fluctuations lasted for about 5 minutes and then stopped. Does this mean that files which needs to be compacted was not found?

Then I tried to call collection.compact() for another larger collection. It lasted for about 80 minutes, and the CPU usage was quite high. The entity count did decrease significantly. 
However, I encountered "compact wrong" errors again, which interrupted the compact process. It's unfortunate.

datanode [2023/07/25 07:16:44.018 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861430567254] [error="upload to blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.com/m  
ilvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"]                                                    
datanode [2023/07/25 07:20:15.654 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861430567255] [error="upload to blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.com/m  
ilvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"]                                                    
datanode [2023/07/25 07:21:52.884 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861430567256] [error="upload to blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.com/m  
ilvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"]                                                    
datanode [2023/07/25 07:25:23.013 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861430567257] [error="upload to blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.com/m  
ilvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"]

**Comment by xiaofan-luan** [comment]
> @xiaofan-luan I haven't installed Prometheus yet, so I can only rely on basic server monitoring. I have made a few calls and noticed some fluctuations in CPU and disk IO, but they were not very high. The fluctuations lasted for about 5 minutes and then stopped. Does this mean that files which needs to be compacted was not found?
> 
> Then I tried to call collection.compact() for another larger collection. It lasted for about 80 minutes, and the CPU usage was quite high. The entity count did decrease significantly. However, I encountered "compact wrong" errors again, which interrupted the compact process. It's unfortunate.
> 
> datanode [2023/07/25 07:16:44.018 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861430567254] [error="upload to blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.com/m ilvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/25 07:20:15.654 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861430567255] [error="upload to blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.com/m ilvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/25 07:21:52.884 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861430567256] [error="upload to blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.com/m ilvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"] datanode [2023/07/25 07:25:23.013 +00:00] [ERROR] [datanode/compactor.go:643] ["compact wrong"] [planID=442909861430567257] [error="upload to blob storage wrong"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.com/m ilvus-io/milvus/internal/datanode/compactor.go:643\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:93"]

the compact wrong did offer any detailed errors. but it should be in your milvus log.

My guess is that your minio again failed to upload. you will need check why minio/S3 keep reporting error. Compaction concurrency is actually limited, and shouldn't hit the object storage quota limitation

**Comment by bedivere2018** [comment]
@xiaofan-luan 
I found a way to obtain the CompactionState and wrote a simple script. The script first calls a compaction and then continuously checks the state. 
```
test_compaction = Collection("test")
test_compaction.compact()
while True:
    print(test_compaction.get_compaction_state())
    time.sleep(10)
```
        

However, I have discovered that there is one compaction ID that remains in the "executing" state. This state has not changed since yesterday and seems to be stuck.
```
CompactionState
 - compaction id: 442909861436321422
 - State: 1
 - executing plan number: 24
 - timeout plan number: 0
 - complete plan number: 0
```

I performed the same operation on the collection by calling "compact" again and checking the state. 
This time, the compaction process completed quickly. However, it seems that the segments that needed compaction were not actually processed.
```
[2023-07-26 11:36:08]
CompactionState
 - compaction id: 442909861436388372
 - State: 1
 - executing plan number: 2
 - timeout plan number: 0
 - complete plan number: 0

[2023-07-26 11:36:28]
CompactionState
 - compaction id: 442909861436388372
 - State: 2
 - executing plan number: 0
 - timeout plan number: 0
 - complete plan number: 2
 ```

Is there any way to let the stuck compaction continue its normal execution, or end it and restart the compaction?
And do you know how to retrieve all the compaction IDs?

**Comment by jiaoew1991** [comment]
Hi @bedivere2018 which version are you using? we released 2.2.12 yesterday which fixed some issues about stuck compaction. You can try it.

**Comment by bedivere2018** [comment]
> Hi @bedivere2018 which version are you using? we released 2.2.12 yesterday which fixed some issues about stuck compaction. You can try it.

@jiaoew1991 
milvus 2.2.11
I saw that 2.2.12 has been released, but it seems that the helm chart has not been updated yet. 
Additionally, I noticed in the release notes for version 2.2.12 that a token must be set when accessing Milvus through RESTful APIs. 
If the server does not have authentication enabled, can any token be used, or is it necessary to use root:Milvus?
